Kernel fusing and stream synchronizes

ThomasHowarth opened the pull request:
**This is a pretty extensive PR, which may take a fair amount of time to review.**
It attempts to make everything consistent and make (micro-)optimisations where possible.
I'll start with a draft as I continue working on it. Here's a preliminary list of changes:
- `std::endl;` -> `\n`(#562)
- `for` loops: `i++`->`++i` (#562)
- lots and lots (and lots) of `const` and `constexpr`(#562)
- Removal of the `PeleLM::` namespace in several places (particularly `PeleLM::TimeStamp`) (#562)
- Consistent passing by value for `int`, `Real`,`TimeStamp` (and other small types, unless they are modified), and by reference for anything in an `amrex::Vector`, or large types. (#562)
- Fusing kernels with the more modern `amrex::ParallelFor(MultiFab...` where possible. There is more to be done here, but I'm not sure if an overload/workaround exists for working on `nodaltilebox`'s. I also need a little guidance to determine exactly when I can `amrex::Gpu::streamSynchronize()` in or outside of `lev` loops.
- Global capture (i.e. `[=]`) has been replaced with explicit capturing (#562, mostly)
- Options that are not local (e.g. `m_incompressible`) are pulled outside of `ParallelFor` loops where possible. (#562)
- Using `reserve` and `emplace_back` where possible, rather than `resize` and `define`, to construct `MultiFab`s in place. (#562)
- Lots of other micro-optimisations/code simplifications. (#562)

If any additional consistent patterns emerge, I'll add them here.

Commit:
first fusing

Commit:
advection final

Commit:
advance and main

Commit:
BC

Commit:
BCfill, BPatch

Commit:
some more changes

Commit:
emplace and diffusion

Commit:
kernel funcs

Commit:
formatting

Commit:
formatting

Commit:
CI fixes

Commit:
dummy mf

Commit:
use species_bc

Commit:
wrong memory, remove dummy

Commit:
formatting

Commit:
naming conv

Commit:
species_bc

Commit:
implicit conversion

Commit:
forces update

Commit:
clang-tidy performance

Commit:
const checker

Commit:
latest changes

Commit:
projection

Commit:
formatting + endl

Commit:
endl

Commit:
header fixes

Commit:
loops

Commit:
advection

Commit:
couple of changes

Commit:
merge and format

Commit:
fixes

ThomasHowarth committed:
Merge branch 'development' into kernel_fusing

Commit:
eb and plasma fixes

Commit:
small change

Commit:
more tweaking and fixes

Commit:
more fixes

Commit:
increment fix

Commit:
fixes

Commit:
typo

Commit:
tolerance tweak

Commit:
merge

Commit:
formatting etc

Commit:
efield

Commit:
efield and extras

Commit:
more fusing

Commit:
typo

Commit:
duplicate + emplace

Commit:
no [=]

Commit:
missing capture

Commit:
indexing

Commit:
rogue space

Commit:
formatting and capture

Commit:
more const and duplicates

Commit:
more emplacing

Commit:
missing capture, consts

Commit:
bot fix

Commit:
constexpr in lambda

Commit:
more consts, plus switch

Commit:
more loops

jrood-nrel commented:
Wow, this all looks great. Thanks for doing all this. I looked over most of it. I have two comments:

- I mostly don't like explicit lambda capturing for maintenance reasons, where it feels tedious for the programmer to find and list everything to capture in `[]` as sometimes they get very long. Does `cppcoreguidelines-avoid-capture-default` allow `clang-tidy --fix` to automatically create the capture list? I'm 50-50 on accepting the drawbacks of `[=]` for the sake of simplicity. If `clang-tidy` can check this for us, then I'm fine doing it explicitly. If it can't, I might argue we don't put the burden on the developer.
- My other comment is that the fused `ParallelFor` changes will take a few more eyes on them as they are delicate, as well as placement for the synchronizes, but it's great if we can fuse the `MFIter` loops everywhere possible.

ThomasHowarth commented:
I don't think `cppcoreguidelines-avoid-capture-default` automatically generates a capture list. I've explicitly changed them for all of the existing lambdas, so I'd hope that it wouldn't be too much of a burden if anyone adds new ones. 

I also agree on the second point - happy for this to take a while to go through as some more eyes look over the kernel fusions. I'll keep looking for any more that I can change easily, but here are some I'm struggling with:
- Any that use `nodaltilebox`s (e.g. [here](https://github.com/AMReX-Combustion/PeleLMeX/blob/f82f7fa75b496ea0d556dab5e466fa6133b05f1c/Source/PeleLMeX_Diffusion.cpp#L294))
- Any that use `HydroUtils::` functions (e.g. [here](https://github.com/AMReX-Combustion/PeleLMeX/blob/f82f7fa75b496ea0d556dab5e466fa6133b05f1c/Source/PeleLMeX_Advection.cpp#L485)), as they take the `mfi` as an argument
- Any that checks whether a box is covered, regular, or contains cut cells for EB  (e.g. [here](https://github.com/AMReX-Combustion/PeleLMeX/blob/f82f7fa75b496ea0d556dab5e466fa6133b05f1c/Source/PeleLMeX_Utils.cpp#L229)). I know that I can check box properties during the fused loop, but I may as well do cell-wise checks there, and that introduces a whole bunch more unnecessary if-else in regular and covered boxes.
- Any that takes a `growntilebox` as part of a function input (e.g. [here](https://github.com/AMReX-Combustion/PeleLMeX/blob/f82f7fa75b496ea0d556dab5e466fa6133b05f1c/Source/PeleLMeX_Soot.cpp#L48))

If anyone has suggestions for any of these, I'll happily take them.

ThomasHowarth commented:
In terms of a demonstration of concrete improvement (beyond consistency within the code itself), three areas seem to do significantly fewer memory allocations, namely `advance::scalars_adv`, `getVelForces` and `differentialDiffusionUpdate`
Current development LMeX:
<img width="1091" height="318" alt="image" src="https://github.com/user-attachments/assets/88e8a773-9484-4fe3-ad7d-8fd41baa1545" />
After these changes:
<img width="1082" height="319" alt="image" src="https://github.com/user-attachments/assets/133236e2-c33e-476e-8a4d-cb2c80de012e" />
There is no significant difference in speed, however, I've not tested on GPUs, or on a large case.

Commit:
better streamsyncs

Commit:
remove comments

baperry2 commented:
Thanks for this @ThomasHowarth. Most of the changes are simple things that I don't worry about at all and make the code much nicer in the long run.

I have no particular opinion on the lambda captures - either way is fine with me but consistency is best. 

I agree with both of you that the major concern here is the StreamSynchronize for fused ParallelFors, as if we mess that up it could cause potentially confusing errors and we don't have an easy way to test for that as it would require many GPU runs. A conservative approach would just be to ensure that each new fused ParallelFor is followed by a StreamSynchronize (@jrood-nrel correct me if I'm wrong but I think that would make them equivalent to the current MFIter + ParallelFor setup). This may limit potential performance benefits, but we could relax it later with further testing in smaller PRs.

Commit:
combine kernels, const&

Commit:
wrong incomp

Commit:
fix typos

Commit:
incomp fixes

Commit:
shift syncs inside

Commit:
much more const, tweaks

Commit:
compile const, endl clang-tidy

ThomasHowarth commented:
Ok, I think I'm done fiddling. I think I've moved all of the `streamSynchonize`s that I've added directly after the `ParallelFor`s - any that I think can be moved, I've left a comment `// Shift outside?` above, and they can be assessed gradually. It would be great if someone could check over the kernel fusions to make sure I haven't broken something.

Commit:
remove endl checks

jrood-nrel commented:
I've decided I'm fine with the explicit lambda captures since I'm seeing that is recommended everywhere I'm looking.

I think we should have @marchdf look at the `ParallelFor` stuff as well. Once we're all fine with the placement of the synchronizes (which I think we're just verifying they're all done immediately to be safe for now), I'm fine merging it.

jrood-nrel commented on Source/PeleLMeX_Forces.cpp:
```diff
@@ -376,8 +366,8 @@ PeleLM::addScalarVarianceSources(const TimeStamp& a_timestamp)
         }
-        amrex::Gpu::streamSynchronize();
       }
+      amrex::Gpu::streamSynchronize();
```
Tagging this one so I can look more at it later.

jrood-nrel replied:
This one should be moved.

jrood-nrel commented on Source/PeleLMeX_ODEQty.cpp:
```diff
@@ -7,21 +7,21 @@ PeleLM::predictODEQty()
       });
-    amrex::Gpu::streamSynchronize();
   }
+  amrex::Gpu::streamSynchronize();
```
Unsure of this one at the moment.

jrood-nrel replied:
This one should be moved.

jrood-nrel commented on Source/Plasma/PeleLMeX_EFTransport.cpp:
```diff
@@ -2,32 +2,32 @@
+        getDiffE(i, j, k, factor, T, mobE_ma[box_no], diffE_ma[box_no]);
+      });
   }
+  amrex::Gpu::streamSynchronize();
```
Tagging this one.

jrood-nrel replied:
This one should be moved.

marchdf commented on Source/PeleLMeX_Advection.cpp:
```diff
@@ -238,79 +242,80 @@ PeleLM::updateVelocity(std::unique_ptr<AdvanceAdvData>& advData)
+      [state_old_ma, adv_aofs_ma, force_ma, state_new_ma,
+       dt_loc =
+         m_dt] AMREX_GPU_DEVICE(int box_no, int i, int j, int k) noexcept {
+        for (int n = 0; n < AMREX_SPACEDIM; ++n) {
```
could add an extra level of parallelism here with int n

marchdf commented on Source/PeleLMeX_Advection.cpp:
```diff
@@ -1108,29 +1119,32 @@ PeleLM::computePassiveAdvTerms(
+      ldataOld_p->state,
+      [state_old_ma, adv_aofs_ma, ext_ma, state_new_ma, state_comp, ncomp,
+       dt = m_dt] AMREX_GPU_DEVICE(int box_no, int i, int j, int k) noexcept {
+        for (int n = state_comp; n < state_comp + ncomp; ++n) {
```
same comment here for increasing the paralellism

marchdf commented on Source/PeleLMeX_Advection.cpp:
```diff
@@ -1108,29 +1119,32 @@ PeleLM::computePassiveAdvTerms(
+        }
+      });
+    // Shift outside?
+    amrex::Gpu::streamSynchronize();
```
yeah you can shift outside. But I get wanting to be cautious.

ThomasHowarth replied:
Let me know if you think some of the others can be shifted outside (all labelled with this comment)

marchdf commented on Source/PeleLMeX_Diffusion.cpp:
```diff
@@ -1100,34 +1124,32 @@ PeleLM::differentialDiffusionUpdate(
+    auto const& fAux_ma =
+      (m_nAux > 0) ? advData->Forcing_aux[lev].arrays() : fY_ma;
+
+    amrex::ParallelFor(
```
I would break this parallelfor in 2 and then increase the parallelism on numspec and numaux

marchdf commented on Source/PeleLMeX_Diffusion.cpp:
```diff
@@ -1299,49 +1323,92 @@ PeleLM::differentialDiffusionUpdate(
-            if (use_soret != 0) {
-              rhoY(i, j, k, n) -= dt * dT(i, j, k, n);
-            }
+        ldata_p->state,
```
Same comment for breaking up and increasing parallelism. You don't need to do these now btw. Another PR is probably better

marchdf commented on Source/PeleLMeX_DiffusionOp.cpp:
```diff
@@ -361,47 +358,48 @@ DiffusionOp::diffuse_scalar(
-            phi_arr(i, j, k, n) = a_phi_arr(i, j, k, n);
+        phi[lev], phi[lev].nGrowVect(),
+        [a_phi_ma, a_rho_ma, phi_ma, phi_comp,
+         ncomp] AMREX_GPU_DEVICE(int box_no, int i, int j, int k) noexcept {
```
also increase parallelism here.

marchdf commented on Source/PeleLMeX_DiffusionOp.cpp:
```diff
@@ -494,27 +492,25 @@ DiffusionOp::diffuse_scalar(
-          } else {
-            a_phi_arr(i, j, k, n) = phi_arr(i, j, k, n);
+        phi[lev], [a_phi_ma, a_rho_ma, phi_ma, ncomp] AMREX_GPU_DEVICE(
+                    int box_no, int i, int j, int k) noexcept {
```
ditto. Lol this should be fun. Might get actual speedup with all of these

marchdf commented on Source/PeleLMeX_DiffusionOp.cpp:
```diff
@@ -618,38 +615,38 @@ DiffusionOp::computeDiffFluxes(
-            phi_arr(i, j, k, n) = a_phi_arr(i, j, k, n);
+        phi[lev], phi[lev].nGrowVect(),
+        [a_phi_ma, a_rho_ma, phi_ma, phi_comp,
+         ncomp] AMREX_GPU_DEVICE(int box_no, int i, int j, int k) noexcept {
```
I might stop noting these places of increased parallelism.

marchdf commented on Source/PeleLMeX_DiffusionOp.cpp:
```diff
@@ -761,38 +758,38 @@ DiffusionOp::computeDiffFluxes(
+         ncomp] AMREX_GPU_DEVICE(int box_no, int i, int j, int k) noexcept {
+          amrex::Array4<amrex::Real const> a_phi_arr(
+            a_phi_ma[box_no], phi_comp);
+          for (int n = 0; n < ncomp; ++n) {
```
yeah going to stop now.

marchdf commented on Source/PeleLMeX_DiffusionOp.cpp:
```diff
@@ -1190,22 +1192,17 @@ DiffusionTensorOp::compute_divtau(
+      amrex::ParallelFor(
+        *a_divtau[lev], [divtau_ma, rho_ma] AMREX_GPU_DEVICE(
+                          int box_no, int i, int j, int k) noexcept {
+          for (int n = 0; n < AMREX_SPACEDIM; ++n) {
```
or will I?

baperry2 commented on Source/PeleLMeX_Forces.cpp:
```diff
@@ -47,201 +49,180 @@ PeleLM::getVelForces(
+      amrex::Array4<amrex::Real const> temp(state_ma[box_no], TEMP);
+      amrex::Array4<amrex::Real const> phiV(state_ma[box_no], PHIV);
+      amrex::Array4<amrex::Real const> nE(state_ma[box_no], NE);
+      amrex::GpuArray<int, 3> blo = ba[box_no].loVect3d();
```
BoxArray is not a device-safe class so this fails on GPU

ThomasHowarth replied:
Merged the plasma stuff into MFIter loop for turb forcing underneath - let me know if it looks good.

baperry2 replied:
I think that should be ok. It compiles fine on GPU now at least. But it looks like another of the recent changes you made is causing the FlameSheetIons case to fail even on CPU.

ThomasHowarth replied:
It's not failing for me with the RhoH fix in PelePhysics - all of the cases fail without that fix. How does yours fail?

baperry2 replied:
Oh, it works with the updated PelePhysics - you can commit the updated PelePhysics here as it should be in development now. 

I do see non-negligible diffs but IDK if that is to be expected. I also see O(1e-9) diffs for the regular FlameSheet case, which is in the gray area of being small but not quite dismissible as a precision issue.

Commit:
streamsync moves, n loops, no ba

ThomasHowarth committed:
Merge branch 'development' into kernel_fusing

Commit:
fix indentations

Commit:
unnecessary header, tol test

Commit:
typo

Commit:
ebbfs tols

ThomasHowarth commented:
One thing is that I've had to change the MLMG tolerances for the EBBFS case in the testing. I know we've seen this particular case be sensitive to small changes, but this still seems a little odd.

ThomasHowarth commented on Source/PeleLMeX_TransportProp.cpp:
```diff
@@ -510,7 +475,8 @@ PeleLM::getDiffusivity(
         const amrex::Box ebx = mfi.tilebox();
```
Is this correct? Seems a little odd to just loop over the tilebox for looking at stuff in the boundaries

Commit:
couple more things

Commit:
simplify

Commit:
ngrow

ThomasHowarth commented on Source/PeleLMeX_Projection.cpp:
```diff
@@ -16,67 +16,73 @@ PeleLM::initialProjection()
         rhs_cc[lev], m_leveldata_new[lev]->divu, 0, 0, 1,
         m_leveldata_new[lev]->divu.nGrow());
       if (m_closed_chamber != 0) {
         rhs_cc[lev].plus(-Sbar, 0, 1);
```
Should there be an ngrow in here, like below?

Commit:
reaction fusing

Commit:
fixes

Commit:
unnecessary capture

Commit:
bump pelephysics

ThomasHowarth commented:
I'm going to shift this to a draft and open another pull request which covers the non-invasive changes (types and consistency stuff), just to split this PR up a bit.

baperry2 commented:
I think the kernel fusing/stream synchronizes are the main thing that needs to be verified. Pretty much everything else I think should be fine given that the tests are passing. So I think we could probably just merge a PR with everything but the ParallelFor fusing changes (which are good, just want to make sure nothing weird happens on GPU).

Commit:
fix conflict

Commit:
cmake test

Commit:
ignore nans in ghost cells for testing

Commit:
duplicate

Commit:
nans in ghost cells

Commit:
atomic adds

Commit:
unnecessary capture

Commit:
some more parallelism

ThomasHowarth committed:
Merge branch 'development' into kernel_fusing

Commit:
typo

ThomasHowarth committed:
Merge branch 'development' into kernel_fusing

ThomasHowarth commented on Source/PeleLMeX_DeriveFunc.cpp:
```diff
@@ -72,19 +72,24 @@ pelelmex_derheatrelease(
+    bx, NUM_SPECIES,
+    [HRR, Hi, react] AMREX_GPU_DEVICE(int i, int j, int k, int n) noexcept {
+      amrex::Real val = -Hi(i, j, k, n) * react(i, j, k, n);
+      amrex::Gpu::Atomic::Add(&HRR(i, j, k), val);
```
Let me know if this is correct - I'm assuming atomic addition is required, but I'm not 100% sure.

baperry2 replied:
I believe atomic adds are required here as well. I think I understand why you split this in two, but is it worth it? Adds an extra streamsync which I'm pretty sure is required and IMO this is a bit less readable.

ThomasHowarth replied:
Fair enough - I'll recombine

jrood-nrel commented:
I talked about this with @marchdf today. I'm hoping to set up an automated test script where we can plug each of these file changes in 1 by 1 and quantify the diffs in some kind of plot since we expect the results to be nondeterministic even without the changes. I'm going to work on trying to set up some testing for this because I would like to get these changes in but we want to be careful.

jrood-nrel committed:
Merge branch 'development' into kernel_fusing

jrood-nrel commented:
Ok I wrote a script that checked the changes in each file one at a time and had it compare against reference plot files for the flamesheet case and an EB case. I also ran the cases with no changes twice to understand the inherent non determinism. Here is a few of the main variables all catted that show the absolute error (on the left) are all not far from machine precision so I think I'm satisfied.
```
 x_velocity                                       0                         0
 y_velocity                                       0                         0
 z_velocity                                       0                         0
 density                                          0                         0
 x_velocity                                       0                         0
 y_velocity                                       0                         0
 z_velocity                                       0                         0
 density                                          0                         0
 x_velocity                                       0                         0
 y_velocity                                       0                         0
 z_velocity                                       0                         0
 density                                          0                         0
 x_velocity                                       0                         0
 y_velocity                                       0                         0
 z_velocity                                       0                         0
 density                                          0                         0
 x_velocity                         2.753353101e-14           3.924120597e-15
 y_velocity                         7.993605777e-15           2.344838948e-15
 z_velocity                         2.888299106e-15            0.001072104352
 density                            7.549516567e-15           5.147540068e-15
 x_velocity                         7.460698725e-14           9.827722554e-15
 y_velocity                         1.776356839e-14           4.163458001e-15
 z_velocity                         6.996735778e-15            0.001245946531
 density                            2.087219286e-14           1.423143431e-14
 x_velocity                         5.217018224e-15           1.807580086e-14
 y_velocity                         6.806190946e-15           2.245464095e-14
 z_velocity                         1.132427485e-14           1.308674338e-14
 density                            3.108624469e-15           2.788503014e-15
 x_velocity                         1.608793394e-14           4.385254253e-14
 y_velocity                         1.812113827e-14           5.050365669e-14
 z_velocity                         2.742250871e-14           3.077952812e-14
 density                            2.664535259e-15           2.389904308e-15
 x_velocity                          5.41788836e-14           7.721656659e-15
 y_velocity                         1.643130076e-14           4.819946727e-15
 z_velocity                         2.148005881e-15           0.0007977283194
 density                            1.421085472e-14           9.689487188e-15
 x_velocity                         1.350031198e-13           1.778349796e-14
 y_velocity                          4.04121181e-14           9.471866952e-15
 z_velocity                         1.063935007e-14            0.001894771897
 density                            3.264055692e-14           2.225554088e-14
 x_velocity                         3.450039733e-15           1.195361575e-14
 y_velocity                         5.113847917e-15           1.687134842e-14
 z_velocity                         8.659739592e-15           1.000750964e-14
 density                            1.998401444e-15           1.792609081e-15
 x_velocity                         1.187570008e-14           3.237082181e-14
 y_velocity                         1.424853074e-14           3.971069003e-14
 z_velocity                         2.398081733e-14           2.691651042e-14
 density                            2.886579864e-15           2.589063001e-15
 x_velocity                          1.33226763e-14           1.898768031e-15
 y_velocity                         5.107025913e-15            1.49809155e-15
 z_velocity                          2.18772152e-15           0.0008119462849
 density                            5.107025913e-15           3.482159458e-15
 x_velocity                         5.950795412e-14           7.838778704e-15
 y_velocity                         1.287858709e-14           3.018507051e-15
 z_velocity                         6.089509513e-15            0.001084458006
 density                            1.976196984e-14           1.347444312e-14
 x_velocity                         3.307141887e-15           1.145850668e-14
 y_velocity                         4.737719548e-15           1.563044473e-14
 z_velocity                         7.771561172e-15           8.981098398e-15
 density                            2.664535259e-15           2.390145441e-15
 x_velocity                         1.217255463e-14           3.317998892e-14
 y_velocity                         1.554981899e-14           4.333738356e-14
 z_velocity                         2.053912596e-14           2.305349272e-14
 density                            2.442490654e-15           2.190745616e-15
 x_velocity                         1.953992523e-14           2.784859779e-15
 y_velocity                         8.881784197e-15           2.605376609e-15
 z_velocity                         2.801130485e-15            0.001040343608
 density                            4.440892099e-15           3.027964746e-15
 x_velocity                         5.595524044e-14           7.370791916e-15
 y_velocity                         1.643130076e-14           3.851198651e-15
 z_velocity                         5.869397026e-15            0.001045194413
 density                            1.154631946e-14            7.87270834e-15
 x_velocity                         3.119574911e-15           1.080862908e-14
 y_velocity                         5.472347835e-15           1.805409322e-14
 z_velocity                         7.882583475e-15           9.109399803e-15
 density                            2.442490654e-15           2.190966654e-15
 x_velocity                         1.206402599e-14            3.28841612e-14
 y_velocity                         1.696129944e-14           4.727118304e-14
 z_velocity                         2.298161661e-14           2.579498915e-14
 density                            3.108624469e-15           2.788221693e-15
 x_velocity                         1.509903313e-14           2.151937102e-15
 y_velocity                         3.774758284e-15           1.107285059e-15
 z_velocity                         2.133743172e-15           0.0007919078089
 density                            4.662936703e-15           3.179362983e-15
 x_velocity                         6.394884622e-14           8.423762189e-15
 y_velocity                         1.065814104e-14           2.498074801e-15
 z_velocity                           6.7712925e-15            0.001205785866
 density                            1.998401444e-14           1.362584136e-14
 x_velocity                         4.166263688e-15            1.44351715e-14
 y_velocity                         5.432979438e-15           1.792421099e-14
 z_velocity                         9.658940314e-15           1.116222229e-14
 density                            2.664535259e-15           2.390145441e-15
 x_velocity                         1.587749029e-14           4.327891457e-14
 y_velocity                         2.052584448e-14           5.720557878e-14
 z_velocity                         2.997602166e-14           3.364563802e-14
 density                            3.108624469e-15           2.788221693e-15
 x_velocity                         1.687538997e-14           2.405106172e-15
 y_velocity                         3.996802889e-15           1.172419474e-15
 z_velocity                         2.792042083e-15            0.001036121686
 density                            4.662936703e-15           3.179362983e-15
 x_velocity                         6.394884622e-14           8.423762189e-15
 y_velocity                         1.065814104e-14           2.498074801e-15
 z_velocity                         7.985377386e-15            0.001422087608
 density                            2.065014826e-14           1.408003607e-14
 x_velocity                         4.633337984e-15           1.605347943e-14
 y_velocity                         5.869951877e-15           1.936584836e-14
 z_velocity                         1.076916334e-14           1.244523635e-14
 density                            2.664535259e-15           2.390145441e-15
 x_velocity                         1.581525709e-14            4.31092791e-14
 y_velocity                         2.442547575e-14           6.807386066e-14
 z_velocity                          3.14193116e-14           3.526561319e-14
 density                            2.664535259e-15           2.389904308e-15
 x_velocity                         1.154631946e-14            1.64559896e-15
 y_velocity                         3.497202528e-15            1.02586704e-15
 z_velocity                         2.245761516e-15           0.0008336199059
 density                            3.996802889e-15           2.725168272e-15
 x_velocity                          6.30606678e-14           8.306765492e-15
 y_velocity                         2.442490654e-14           5.724754751e-15
 z_velocity                         6.928475925e-15            0.001233928907
 density                            2.042810365e-14           1.392863783e-14
 x_velocity                         5.155598171e-15           1.786299412e-14
 y_velocity                         7.267624003e-15           2.397697755e-14
 z_velocity                         9.880984919e-15           1.141882511e-14
 density                            1.998401444e-15           1.792609081e-15
 x_velocity                         1.601171452e-14           4.364478341e-14
 y_velocity                         2.333587967e-14           6.503715373e-14
 z_velocity                         3.186340081e-14           3.576406708e-14
 density                            2.886579864e-15           2.589063001e-15
 x_velocity                         3.641531521e-14           5.189965951e-15
 y_velocity                         1.065814104e-14           3.126451931e-15
 z_velocity                         2.361892463e-15           0.0008766049183
 density                            1.088018564e-14           7.418513628e-15
 x_velocity                         8.615330671e-14           1.134867962e-14
 y_velocity                         2.042810365e-14           4.787976701e-15
 z_velocity                         9.271752268e-15             0.00165114229
 density                            2.176037128e-14           1.483702726e-14
 x_velocity                         4.341470759e-15            1.50422248e-14
 y_velocity                         6.141774914e-15           2.026263318e-14
 z_velocity                         1.099120794e-14           1.270183916e-14
 density                            1.998401444e-15           1.792609081e-15
 x_velocity                         1.707206425e-14           4.653508813e-14
 y_velocity                         2.159764609e-14           6.019269249e-14
 z_velocity                         3.053113318e-14           3.426870539e-14
 density                            3.330669074e-15           2.987380385e-15
 x_velocity                          5.41788836e-14           7.721656659e-15
 y_velocity                         1.509903313e-14           4.429140235e-15
 z_velocity                         2.767343465e-15            0.001026965487
 density                            1.265654248e-14           8.629699527e-15
 x_velocity                         1.358912982e-13           1.790049465e-14
 y_velocity                           3.5971226e-14           8.431002452e-15
 z_velocity                         1.076910238e-14            0.001917871727
 density                            3.108624469e-14           2.119575322e-14
 x_velocity                         3.292938838e-15            1.14092963e-14
 y_velocity                         5.074011957e-15           1.673992363e-14
 z_velocity                          8.54871729e-15           9.879208237e-15
 density                            2.664535259e-15           2.390145441e-15
 x_velocity                         1.292260569e-14           3.522448054e-14
 y_velocity                         1.625582264e-14           4.530501747e-14
 z_velocity                         2.475797345e-14           2.778880474e-14
 density                            2.886579864e-15           2.589063001e-15
 x_velocity                         2.575717417e-14           3.670951526e-15
 y_velocity                         7.105427358e-15           2.084301287e-15
 z_velocity                          2.30838752e-15           0.0008567257923
 density                            5.551115123e-15           3.784955933e-15
 x_velocity                         6.217248938e-14           8.189768795e-15
 y_velocity                         1.643130076e-14           3.851198651e-15
 z_velocity                         6.505472208e-15            0.001158568658
 density                            2.131628207e-14           1.453423078e-14
 x_velocity                         3.908982513e-15           1.354374979e-14
 y_velocity                         5.288060571e-15             1.7446102e-14
 z_velocity                         7.993605777e-15           9.237701209e-15
 density                            3.108624469e-15           2.788503014e-15
 x_velocity                         1.274425444e-14            3.47383301e-14
 y_velocity                         1.639703998e-14           4.569859053e-14
 z_velocity                         2.298161661e-14           2.579498915e-14
 density                            2.886579864e-15           2.589063001e-15
 x_velocity                         3.197442311e-14           4.557043274e-15
 y_velocity                         9.325873407e-15            2.73564544e-15
 z_velocity                         2.557646086e-15           0.0009494356442
 density                            9.992007222e-15           6.812920679e-15
 x_velocity                         8.082423619e-14           1.064669943e-14
 y_velocity                         2.220446049e-14           5.204322501e-15
 z_velocity                         7.426920758e-15             0.00132262767
 density                             2.24265051e-14           1.529122197e-14
 x_velocity                          4.96087546e-15            1.71883235e-14
 y_velocity                          5.74214477e-15           1.894419362e-14
 z_velocity                         9.769962617e-15            1.12905237e-14
 density                            3.108624469e-15           2.788503014e-15
 x_velocity                         1.589288597e-14           4.332088014e-14
 y_velocity                         2.478989304e-14           6.908949255e-14
 z_velocity                         2.986499936e-14           3.352102455e-14
 density                            2.886579864e-15           2.589063001e-15
 x_velocity                         2.842170943e-14           4.050705132e-15
 y_velocity                         9.325873407e-15            2.73564544e-15
 z_velocity                         1.719649791e-15           0.0006386880725
 density                            9.547918012e-15           6.510124204e-15
 x_velocity                         8.260059303e-14           1.088069283e-14
 y_velocity                          2.26485497e-14           5.308408951e-15
 z_velocity                         7.650887193e-15            0.001362488885
 density                             2.24265051e-14           1.529122197e-14
 x_velocity                         4.769513777e-15           1.652529808e-14
 y_velocity                         6.841708732e-15           2.257181943e-14
 z_velocity                         1.310063169e-14           1.513956587e-14
 density                            3.108624469e-15           2.788503014e-15
 x_velocity                         1.817698485e-14           4.954688428e-14
 y_velocity                         1.959313246e-14           5.460610809e-14
 z_velocity                         3.264055692e-14            3.66363614e-14
 density                            2.664535259e-15           2.389904308e-15
 x_velocity                         3.819167205e-14           5.443135022e-15
 y_velocity                         1.154631946e-14           3.386989592e-15
 z_velocity                         1.917120863e-15           0.0007116322192
 density                            8.659739592e-15           5.904531255e-15
 x_velocity                         9.414691249e-14           1.240164989e-14
 y_velocity                         2.797762022e-14           6.557446352e-15
 z_velocity                         6.989017829e-15            0.001244620397
 density                            2.087219286e-14           1.423143431e-14
 x_velocity                         3.247239717e-15           1.125095906e-14
 y_velocity                         4.777265823e-15           1.576091379e-14
 z_velocity                         8.659739592e-15           1.000750964e-14
 density                            1.998401444e-15           1.792609081e-15
 x_velocity                         1.208646898e-14           3.294533637e-14
 y_velocity                         1.665770928e-14           4.642507652e-14
 z_velocity                         2.164934898e-14           2.429962746e-14
 density                            3.108624469e-15           2.788221693e-15
 x_velocity                         2.309263891e-14            3.29119792e-15
 y_velocity                         6.661338148e-15           1.954032457e-15
 z_velocity                         2.918495203e-15            0.001083812972
 density                            4.662936703e-15           3.179362983e-15
 x_velocity                         5.773159728e-14            7.60478531e-15
 y_velocity                         2.042810365e-14           4.787976701e-15
 z_velocity                         4.701996354e-15           0.0008373129132
 density                            1.287858709e-14           8.781097764e-15
 x_velocity                         4.498951125e-15           1.558785903e-14
 y_velocity                         6.352577698e-15           2.095810306e-14
 z_velocity                         1.043609643e-14           1.206033213e-14
 density                            3.108624469e-15           2.788503014e-15
 x_velocity                         1.629241447e-14           4.440991623e-14
 y_velocity                          2.06055164e-14           5.742762461e-14
 z_velocity                         3.375077995e-14           3.788249614e-14
 density                            2.886579864e-15           2.589063001e-15
 x_velocity                         2.309263891e-14            3.29119792e-15
 y_velocity                         6.661338148e-15           1.954032457e-15
 z_velocity                         2.376207034e-15           0.0008818154002
 density                            5.329070518e-15           3.633557695e-15
 x_velocity                         6.394884622e-14           8.423762189e-15
 y_velocity                          1.82076576e-14           4.267544451e-15
 z_velocity                          6.16257195e-15            0.001097509768
 density                            2.087219286e-14           1.423143431e-14
 x_velocity                         3.127923268e-15           1.083755426e-14
 y_velocity                         5.075184251e-15           1.674379121e-14
 z_velocity                         7.993605777e-15           9.237701209e-15
 density                            3.108624469e-15           2.788503014e-15
 x_velocity                         1.387388468e-14           3.781747989e-14
 y_velocity                         1.728483892e-14           4.817288837e-14
 z_velocity                         2.509104036e-14           2.816264516e-14
 density                            2.442490654e-15           2.190745616e-15
 x_velocity                         3.019806627e-14           4.303874203e-15
 y_velocity                         8.881784197e-15           2.605376609e-15
 z_velocity                         2.089250878e-15           0.0007754934396
 density                            7.549516567e-15           5.147540068e-15
 x_velocity                         7.815970093e-14           1.029570934e-14
 y_velocity                          2.26485497e-14           5.308408951e-15
 z_velocity                         6.618129906e-15            0.001178539541
 density                            2.153832668e-14           1.468562902e-14
 x_velocity                         5.218481897e-15           1.808087216e-14
 y_velocity                         7.580673827e-15            2.50097757e-14
 z_velocity                         1.310063169e-14           1.513956587e-14
 density                            2.664535259e-15           2.390145441e-15
 x_velocity                         1.476705043e-14           4.025207398e-14
 y_velocity                         1.928746876e-14           5.375422261e-14
 z_velocity                         2.886579864e-14           3.239950328e-14
 density                            3.330669074e-15           2.987380385e-15
 x_velocity                          1.33226763e-14           1.898768031e-15
 y_velocity                         5.107025913e-15            1.49809155e-15
 z_velocity                         2.841687066e-15            0.001054644682
 density                            4.884981308e-15           3.330761221e-15
 x_velocity                          6.30606678e-14           8.306765492e-15
 y_velocity                          1.33226763e-14           3.122593501e-15
 z_velocity                         6.293269477e-15            0.001120797438
 density                            2.087219286e-14           1.423143431e-14
 x_velocity                         3.382927619e-15           1.172108728e-14
 y_velocity                         5.263483063e-15           1.736501712e-14
 z_velocity                          8.10462808e-15           9.366002615e-15
 density                            3.108624469e-15           2.788503014e-15
 x_velocity                          1.19374996e-14           3.253927515e-14
 y_velocity                         1.763639148e-14            4.91526662e-14
 z_velocity                         2.309263891e-14           2.591960262e-14
 density                            2.664535259e-15           2.389904308e-15
```

jrood-nrel committed:
Merge branch 'development' into kernel_fusing

jrood-nrel commented on Source/PeleLMeX_Utils.cpp:
```diff
@@ -2201,20 +2201,18 @@ PeleLM::extendSignedDistance(
   // First set the region far away at the max value we need
   const auto& dx = geom[0].CellSizeArray();
```
Sorry I messed up the conflict resolution. This line should replace line 2205 I believe.

baperry2 commented:
@jrood-nrel I don't think I understand your output. What are all the different comparisons? Can you include species and temperature as well?

baperry2 commented on Source/PeleLMeX_Plot.cpp:
```diff
@@ -1052,30 +1045,33 @@ PeleLM::initLevelDataFromPlt(int a_lev, const std::string& a_dataPltFile)
+            soot(i, j, k, n) *= std::pow(100., soot_exp[n]);
+          });
+        amrex::Gpu::streamSynchronize();
+        ldata_p->state.mult(1.E6, FIRSTSOOT + NUM_SOOT_MOMENTS - 1, 1);
```
I think this line should have `FIRSTSOOT + NUMSOOTVAR - 1` rather than `FIRSTSOOT + NUM_SOOT_MOMENTS - 1` in order to match the previous implementation

ThomasHowarth replied:
I've fixed, but also just combined it into the ParallelFor above for similar reasons as the HRR one from earlier.

ThomasHowarth committed:
Merge branch 'development' into kernel_fusing

ThomasHowarth committed:
utils fix

ThomasHowarth commented:
Sorry for the delay - I've fixed the issue with the merge. Hopefully this is in good enough shape now... Perhaps @jrood-nrel can just confirm these numbers for @baperry2 ?

I'll open another PR playing around with the position of the synchronizes, but I feel it's best we just get this in ASAP - it's been in the pipeline a while.

jrood-nrel commented:
I gave @baperry2 the output for all the variables. I'm still fine with merging this.

baperry2 commented:
Started going through what Jon sent me a while ago but then got distracted by other stuff, I'll get back on it.

baperry2 commented on Source/Plasma/PeleLMeX_EFIonDrift.cpp:
```diff
@@ -59,45 +59,39 @@ PeleLM::ionDriftVelocity(const std::unique_ptr<AdvanceAdvData>& advData)
+            mob_h_ma[box_no](i, j, k, n) * -0.5 *
+            (gp_o_ma[box_no](i, j, k) + gp_n_ma[box_no](i, j, k));
+        });
+      // Shift outside (idim/lev)?
```
Pretty sure this should be ok for all idim/lev loops, but that should come in a subsequent PR.

baperry2 commented on Source/Plasma/PeleLMeX_EFPoisson.cpp:
```diff
@@ -25,24 +25,23 @@ PeleLM::poissonSolveEF(const TimeStamp a_time)
-          }
-        });
-    }
+    constexpr amrex::Real factor = -1.0;
```
Why break this into two separate ParallelFors?

baperry2 commented on Source/PeleLMeX_Init.cpp:
```diff
@@ -366,26 +361,26 @@ PeleLM::initLevelData(const int lev)
-      (m_nAux > 0) ? ldata_p->auxiliaries.array(mfi) : DummyFab.array();
-    amrex::ParallelFor(bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+  // don't want to use state for dummy in case user overwrites state in aux
+  amrex::MultiFab dummy_mf(grids[lev], dmap[lev], 1, 0);
```
How about setting number of components to 0 in the dummy?

baperry2 commented on Source/PeleLMeX_DiffusionOp.cpp:
```diff
@@ -146,26 +146,23 @@ DiffusionOp::diffuse_scalar(
-          } else {
-            phi_arr(i, j, k, n) = a_phi_arr(i, j, k, n);
-          }
+        phi[lev], phi[lev].nGrowVect(), ncomp,
```
Replace with a Multifab divide call, similar to the multifab copy above? Same for other instances below in this file.

ThomasHowarth replied:
I thought about this, but we would need to copy (just outside of the if), then divide separately, meaning we have two separate ParallelFor's. I figure it's nicer if we just have 1.

baperry2 commented on Source/PeleLMeX_DiffusionOp.cpp:
```diff
@@ -281,27 +278,24 @@ DiffusionOp::diffuse_scalar(
+        phi[lev], amrex::IntVect(0), ncomp,
+        [a_phi_ma, a_rho_ma, phi_ma] AMREX_GPU_DEVICE(
+          int box_no, int i, int j, int k, int n) noexcept {
+          a_phi_ma[box_no](i, j, k, n) =
```
Just use multifab multiply instead?

ThomasHowarth replied:
same here as above

baperry2 commented on Source/PeleLMeX_Forces.cpp:
```diff
@@ -49,129 +49,102 @@ PeleLM::getVelForces(
+      [force_ma, state_ma] AMREX_GPU_DEVICE(
+        int box_no, int i, int j, int k, int n) noexcept {
+        amrex::Array4<amrex::Real const> rho(state_ma[box_no], DENSITY);
+        force_ma[box_no](i, j, k, n) /= rho(i, j, k);
```
could use the multifab divide routine for further simplification

ThomasHowarth replied:
Can't use multifab divide since we're iterating in one multifab, but not the other.

baperry2 commented on Source/PeleLMeX_Reactions.cpp:
```diff
@@ -484,28 +483,28 @@ PeleLM::getHeatRelease(const int a_lev, amrex::MultiFab* a_HR)
+  amrex::MultiFab Enth(grids[a_lev], dmap[a_lev], NUM_SPECIES, 0);
+  auto const& enth_ma = Enth.arrays();
+  amrex::ParallelFor(
+    *a_HR, [state_n_ma, enth_ma, leosparm] AMREX_GPU_DEVICE(
```
again, is it worth splitting this into two separate parallelfors?

baperry2 reviewed (commented):
Just some minor comments, the only thing that definitely needs to be changed before merging is the soot comment in `PeleLMeX_Plot.cpp`

ThomasHowarth committed:
Merge branch 'development' into kernel_fusing

Commit:
combining some parallelfors, soot fix

Commit:
smaller dummy, simplify

Commit:
argument typo

Commit:
undo divide

baperry2 reviewed (approved):
Should be good to go once tests pass. Thanks again for doing this!
